protocol/contracts/nexus/Relic.sol

```diff
@@ -0,0 +1,549 @@
+import { IERC721A } from "../interfaces/nexus/IERC721A.sol";
+import { CommonEventsAndErrors } from "../common/CommonEventsAndErrors.sol";
+import { TempleElevatedAccess } from "../v2/access/TempleElevatedAccess.sol";
+import { IShard } from "../interfaces/nexus/IShard.sol";
+
+
+contract Relic is ERC721ACustom, ERC1155Receiver, TempleElevatedAccess {
+    using SafeERC20 for IERC20;
+    using EnumerableSet for EnumerableSet.UintSet;
+
+    IShard public shard;
+
+    address private constant ZERO_ADDRESS = 0x0000000000000000000000000000000000000000;
+    uint256 private constant PER_MINT_QUANTITY = 0x01;
+    bytes private constant ZERO_BYTES = "";
+
+    /// @notice mapping for information about relic. relicId to RelicInfo
+    mapping(uint256 => RelicInfo) public relicInfos;
+
+    /// @notice base uris for different rarities
+    mapping(Rarity => string) private baseUris;
+
+    /// @notice XP thresholds for Relic rarity levels
+    mapping(Rarity => uint256) public rarityXPThresholds;
+
+    /// @notice to keep track of blacklisted accounts and shard balances
+    mapping(address => mapping(uint256 => uint256)) public blacklistedAccountShards;
+    mapping(address => bool) public blacklistedAccounts;
+    mapping(address => bool) public relicMinters;
+    mapping(address => EnumerableSet.UintSet) private ownerRelics;
+
+    enum Rarity {
+        Common,
+        Uncommon,
+        Rare,
+        Epic,
+        Legendary
+    }
+
+    enum Enclave {
+        Chaos,
+        Mystery,
+        Logic,
+        Order,
+        Structure
+    }
+
+    struct RelicInfo {
+        Enclave enclave;
+        Rarity rarity;
+        uint128 xp;
+        /// @notice shards equipped to this contract. can extract owner of relic from ownerOf(relicId)
+        mapping(uint256 => uint256) equippedShards;
+    }
+
+    event RarityXPThresholdSet(Rarity rarity, uint256 threshold);
+    event RarityBaseUriSet(Rarity rarity, string uri);
+    event RelicMinterSet(address minter, bool allow);
+    event ShardSet(address shard);
+    event RelicMinted(address to, uint256 nextTokenId, uint256 quantity);
+    event RelicXPSet(uint256 relicId, uint256 xp);
+    event ShardEquipped(address caller, uint256 relicId, uint256 shardId, uint256 amount);
+    event ShardsEquipped(address caller, uint256 relicId, uint256[] shardIds, uint256[] amounts);
+    event ShardUnequipped(address caller, uint256 relicId, uint256 shardId, uint256 amount);
+    event ShardsUnequipped(address recipient, uint256 relicId, uint256[] shardIds, uint256[] amounts);
+    event AccountBlacklistSet(address account, bool blacklist, uint256[] shardIds, uint256[] amounts);
+
+    error InvalidParamLength();
+    error CallerCannotMint(address msgSender);
+    error InvalidRelic(uint256 relicId);
+    error InvalidAddress(address invalidAddress);
+    error InsufficientShardBalance(uint256 actualBalance, uint256 requestedBalance);
+    error ZeroAddress();
```

Comment:
Could be put into CommonErrorsAndAlerts? Or just use the existin `CommonErrorsAndAlerts.InvalidAddress()`